sa/model.go

```diff
@@ -64,15 +65,22 @@ func selectRegistration(s db.OneSelector, q string, args ...interface{}) (*regMo
 
 const certFields = "registrationID, serial, digest, der, issued, expires"
 
-// SelectCertificate selects all fields of one certificate object
-// identified by serial.
+// SelectCertificate selects all fields of one certificate object identified by
+// a serial. If more than one certificate contains the same serial only the
+// first is returned (along with a `berrors.DuplicateError`).
 func SelectCertificate(s db.OneSelector, serial string) (core.Certificate, error) {
 	var model core.Certificate
-	err := s.SelectOne(
-		&model,
-		"SELECT "+certFields+" FROM certificates WHERE serial = ?",
-		serial,
-	)
+	query := "SELECT " + certFields + " FROM certificates WHERE serial = ?"
+	err := s.SelectOne(&model, query, serial)
+	if errors.Is(err, fmt.Errorf("gorp: multiple rows returned for: %s - %v", query, serial)) {
+		// Re-attempt but limit our query to just one result.
+		err = s.SelectOne(&model, query+" LIMIT 1", serial)
+		if err != nil {
+			return model, err
+		} else {
+			return model, berrors.DuplicateError("multiple certificates for serial %s", serial)
```

Comment:
I'm having trouble finding a reference right now, but it's a pretty strong convention in Go that you _either_ return a useful value _or_ return a non-nil error, not both. There are a few small exceptions in the stdlib but in general it's a very good rule to follow.

Here's what I think we should do here: Switch from `SelectOne` to `Select` and add a `LIMIT 1`, so things "just work" when there are duplicates. That means we won't have an error log line. That's fine, IMO. If we want to validate the no-duplicates property we should do it with a periodic bulk query rather than something that can be triggered arbitrarily many times by user traffic (or not at all, depending on traffic patterns).

Reply:
(You can use `SelectOne` with a `LIMIT 1`, I believe.)

Reply:
Yeah, the better approach here is just a `SelectOne` with a `LIMIT 1`. `Select` is doesn't return `sql.NoRows` which means the wrapped executor won't produce a `ErrDatabaseOp` wrapping a `sql.NoRows`. This would lead to having to refactor `db.IsNoRows` which currently uses errors.As() to assert the returned `err` to an `ErrDatabaseOp` and call the `IsNoRows()` method.